### PR TITLE
Implement complex.__new__ and complex.__neg__

### PR DESCRIPTION
**Pull request by m4ns0ur**

### CODE REVIEW

File: runtime/complex.go [hunk 1]
```diff
@@ -130,3 +134,130 @@ func complexCoerce(o *Object) (complex128, bool) {
...
+				return nil, f.RaiseType(TypeErrorType, fmt.Sprintf(format, result.typ.Name()))
+			}
+			return result, nil
+		}
+		if floatSlot := o.typ.slots.Float; floatSlot != nil {
+			result, raised := floatConvert(floatSlot, f, o)
+			if raised != nil {
+				return nil, raised
+			}
+			f := toFloatUnsafe(result).Value()
```
Comment (trotterdylan) [inline runtime/complex.go]: If floatConvert() returned (*Float, *BaseException) then this unsafe cast would be unnecessary.

File: runtime/complex.go [hunk 2]
```diff
@@ -130,3 +134,130 @@ func complexCoerce(o *Object) (complex128, bool) {
...
+// If string is well-formed (one of these forms: <float>, <float>j,
+// <float><signed-float>j, <float><sign>j, <sign>j or j, where <float> is
+// any numeric string that's acceptable by strconv.ParseFloat(s, 64)),
+// ParseComplex returns the respective complex128 number.
+func parseComplex(s string) (complex128, error) {
+	c := strings.Count(s, "(")
+	if (c > 1) || (c == 1 && strings.Count(s, ")") != 1) {
+		return complex(0, 0), errors.New("Malformed complex string, more than one matching parantheses")
+	}
+	ts := strings.Trim(s, "() ")
```
Comment (trotterdylan) [inline runtime/complex.go]: Note that all whitespace is allowed around the parentheses:

```
>>> complex(' (\n1+2j\t)  \t ')
(1+2j)
```

So maybe the thing to do is to TrimSpace(s), remove parens from the beginning and end and then do another TrimSpace() on the inside of the parens.

File: runtime/complex.go [hunk 3]
```diff
@@ -130,3 +134,130 @@ func complexCoerce(o *Object) (complex128, bool) {
...
+	sj := `(?P<sign5>[-+])j`
+	j := `(?P<onlyJ>j)`
+	r := regexp.MustCompile(`^(?:` + fsfj + `|` + fsj + `|` + fj + `|` + f + `|` + sj + `|` + j + `)$`)
+	subs := r.FindStringSubmatch(ts)
+	if subs == nil {
+		return complex(0, 0), errors.New("Malformed complex string, no mathing pattern found")
+	}
+	res := make(map[string]string)
+	for i, name := range r.SubexpNames() {
+		// First one is the complete string
```
Comment (trotterdylan) [inline runtime/complex.go]: Comments should be full sentences (including period).

File: runtime/complex.go [hunk 4]
```diff
@@ -130,3 +134,130 @@ func complexCoerce(o *Object) (complex128, bool) {
...
+	fj := `(?P<imag3>` + fre + `)j`
+	f := `(?P<real4>` + fre + `)`
+	sj := `(?P<sign5>[-+])j`
+	j := `(?P<onlyJ>j)`
+	r := regexp.MustCompile(`^(?:` + fsfj + `|` + fsj + `|` + fj + `|` + f + `|` + sj + `|` + j + `)$`)
+	subs := r.FindStringSubmatch(ts)
+	if subs == nil {
+		return complex(0, 0), errors.New("Malformed complex string, no mathing pattern found")
+	}
+	res := make(map[string]string)
```
Comment (trotterdylan) [inline runtime/complex.go]: Allocation in Go is relatively expensive. I think you can avoid allocating this map by using indices for the sub-matches and using a fixed length array: var res [6]string. Constants for the index names would make it more readable.

File: runtime/complex_test.go [hunk 5]
```diff
@@ -80,6 +83,92 @@ func TestComplexNE(t *testing.T) {
...
+	}
+	subType := newTestClass("SubType", []*Type{ComplexType}, newStringDict(map[string]*Object{}))
+	subTypeObject := (&Complex{Object: Object{typ: subType}, value: 3.14}).ToObject()
+	slotSubTypeType := newTestClass("SlotSubType", []*Type{ObjectType}, newStringDict(map[string]*Object{
+		"__complex__": newBuiltinFunction("__complex__", func(_ *Frame, _ Args, _ KWArgs) (*Object, *BaseException) {
+			return subTypeObject, nil
+		}).ToObject(),
+	}))
+	cases := []invokeTestCase{
+		{args: wrapArgs(ComplexType), want: NewComplex(0).ToObject()},
```
Comment (trotterdylan) [inline runtime/complex_test.go]: Love all these test cases! Thanks for being so thorough.

File: runtime/complex.go [hunk 6]
```diff
@@ -130,3 +134,133 @@ func complexCoerce(o *Object) (complex128, bool) {
...
+				return nil, f.RaiseType(TypeErrorType, fmt.Sprintf(format, result.typ.Name()))
+			}
+			return result, nil
+		}
+		if floatSlot := o.typ.slots.Float; floatSlot != nil {
+			result, raised := floatConvert(floatSlot, f, o)
+			if raised != nil {
+				return nil, raised
+			}
+			f := result.Value()
```
Comment (trotterdylan) [inline runtime/complex.go]: I think you can ditch this temporary variable f and pass result.Value() directly into complex() below.

File: runtime/complex.go [hunk 7]
```diff
@@ -130,3 +134,133 @@ func complexCoerce(o *Object) (complex128, bool) {
...
+		}
+		s := toStrUnsafe(o).Value()
+		result, err := parseComplex(s)
+		if err != nil {
+			return nil, f.RaiseType(ValueErrorType, "complex() arg is a malformed string")
+		}
+		return NewComplex(result).ToObject(), nil
+	}
+
+	// TODO: fix me
```
Comment (trotterdylan) [inline runtime/complex.go]: Flagging this TODO to make sure it gets addressed.

File: runtime/complex.go [hunk 8]
```diff
@@ -130,3 +134,133 @@ func complexCoerce(o *Object) (complex128, bool) {
...
+	fj := `(?P<imag3>` + fre + `)j`
+	f := `(?P<real4>` + fre + `)`
+	sj := `(?P<sign5>[-+])j`
+	j := `(?P<onlyJ>j)`
+	r := regexp.MustCompile(`^(?:` + fsfj + `|` + fsj + `|` + fj + `|` + f + `|` + sj + `|` + j + `)$`)
+	subs := r.FindStringSubmatch(ts)
+	if subs == nil {
+		return complex(0, 0), errors.New("Malformed complex string, no mathing pattern found")
+	}
+	const Real1 = 1
```
Comment (trotterdylan) [inline runtime/complex.go]: Nitpick, declare these in a const (...) block. Also, don't capitalize them:

```
const (
    real1 = 1
    imag1 = 2
    ...
)
```

Also, probably a little easier to read this function if these constants are declared at the package scope, rather than in this function.

File: runtime/complex.go [hunk 9]
```diff
@@ -130,3 +134,133 @@ func complexCoerce(o *Object) (complex128, bool) {
...
+		return complex(0, 1), nil
+	}
+	if subs[OnlyJ] != "" {
+		return complex(0, 1), nil
+	}
+	return complex(0, 0), errors.New("Malformed complex string")
+}
+
+func unsignNaN(s string) string {
+	us := strings.ToUpper(s)
```
Comment (trotterdylan) [inline runtime/complex.go]: Nitpick: In this codebase we usually use ToLower() for this kind of normalization.

**Comment by m4ns0ur** [comment]
@trotterdylan I changed the structure based on [CPython Implementation](https://github.com/python/cpython/blob/18b250f844bf8b2d1a81c2d2dcc74e850364fe35/Objects/complexobject.c#L934) to avoid boilerplate
still believe it can be cleaned up, so please advice

File: runtime/complex_test.go [hunk 10]
```diff
@@ -80,6 +83,109 @@ func TestComplexNE(t *testing.T) {
...
+		{args: wrapArgs(ComplexType, math.NaN()), want: NewComplex(complex(math.NaN(), 0)).ToObject()},
+		{args: wrapArgs(ComplexType, math.Inf(1)), want: NewComplex(complex(math.Inf(1), 0)).ToObject()},
+		{args: wrapArgs(ComplexType, math.Inf(-1)), want: NewComplex(complex(math.Inf(-1), 0)).ToObject()},
+		{args: wrapArgs(ComplexType, biggestFloat), want: NewComplex(complex(math.MaxFloat64, 0)).ToObject()},
+		{args: wrapArgs(ComplexType, new(big.Int).Neg(biggestFloat)), want: NewComplex(complex(-math.MaxFloat64, 0)).ToObject()},
+		{args: wrapArgs(ComplexType, new(big.Int).Sub(big.NewInt(-1), biggestFloat)), wantExc: mustCreateException(OverflowErrorType, "long int too large to convert to float")},
+		{args: wrapArgs(ComplexType, new(big.Int).Add(biggestFloat, big.NewInt(1))), wantExc: mustCreateException(OverflowErrorType, "long int too large to convert to float")},
+		{args: wrapArgs(ComplexType, bigLongNumber), wantExc: mustCreateException(OverflowErrorType, "long int too large to convert to float")},
+		// {args: wrapArgs(ComplexType, complex(1, 2)), want: NewComplex(complex(1, 2)).ToObject()},
```
Comment (m4ns0ur) [inline runtime/complex_test.go]: @trotterdylan couldn't pass this test cases yet
something's fishy regarding to Complex Slot

File: runtime/complex_test.go [hunk 11]
```diff
@@ -146,6 +252,61 @@ func TestComplexRepr(t *testing.T) {
...
+		{"((2.1-3.4j))", complex(0, 0), ErrSyntax},
+		{"3.14 -15.16j", complex(0, 0), ErrSyntax},
+		{"3.14- 15.16j", complex(0, 0), ErrSyntax},
+		{"3.14-15.16 j", complex(0, 0), ErrSyntax},
+		{"3.14 - 15.16 j", complex(0, 0), ErrSyntax},
+		{"foo", complex(0, 0), ErrSyntax},
+		{"foo+bar", complex(0, 0), ErrSyntax},
+	}
+	for _, cas := range cases {
+		if got, _ := parseComplex(cas.s); got != cas.want {
```
Comment (trotterdylan) [inline runtime/complex_test.go]: I think you could use complexesAreSame() which was recently added:

```
if got, _ := parseComplex(cas.s); !complexesAreSame(got, cas.want) {
  t.Errorf(...)
}
```

File: runtime/complex.go [hunk 12]
```diff
@@ -169,6 +178,185 @@ func complexCoerce(o *Object) (complex128, bool) {
...
+			return nil, f.RaiseType(ValueErrorType, "complex() arg is a malformed string")
+		}
+		return NewComplex(result).ToObject(), nil
+	}
+	if argc == 2 && args[1].isInstance(StrType) {
+		return nil, f.RaiseType(TypeErrorType, "complex() second arg can't be a string")
+	}
+	var cr complex128
+	crIsComplex := false
+	if complexSlot := o.typ.slots.Complex; complexSlot != nil {
```
Comment (trotterdylan) [inline runtime/complex.go]: Probably worth factoring this `if slots.Complex/else if slots.Float` piece out into a helper to eliminate some duplicate code.

File: runtime/complex.go [hunk 13]
```diff
@@ -169,6 +178,185 @@ func complexCoerce(o *Object) (complex128, bool) {
...
+	} else if floatSlot := args[1].typ.slots.Float; floatSlot != nil {
+		result, raised := floatConvert(floatSlot, f, args[1])
+		if raised != nil {
+			return nil, raised
+		}
+		ci = complex(result.Value(), 0)
+	} else {
+		return nil, f.RaiseType(TypeErrorType, "complex() argument must be a string or a number")
+	}
+	if ciIsComplex {
```
Comment (trotterdylan) [inline runtime/complex.go]: Rather than doing all this bookkeeping, could you just do something like this:

```
cr, raised := complexConvert(args[0])
if raised != nil {
  return nil, raised
}
var ci complex128
if len(args) > 1 {
  if ci, raised = complexConvert(args[1]); raised != nil {
    return nil, raised
  }
}
return NewComplex(cr + ci * 1i).toObject(), nil
```

File: runtime/complex.go [hunk 14]
```diff
@@ -169,6 +178,169 @@ func complexCoerce(o *Object) (complex128, bool) {
...
+		if raised != nil {
+			return complex(0, 0), raised
+		}
+		return complex(result.Value(), 0), nil
+	} else {
+		return complex(0, 0), f.RaiseType(TypeErrorType, "complex() argument must be a string or a number")
+	}
+}
+
+func complexConvert(complexSlot *unaryOpSlot, f *Frame, o *Object) (*Complex, *BaseException) {
```
Comment (trotterdylan) [inline runtime/complex.go]: Please move helper functions to the end of the file. The general order of stuff in these files is: type definition, methods, type init function, helpers.

File: runtime/complex_test.go [hunk 15]
```diff
@@ -132,6 +133,111 @@ func TestComplexNE(t *testing.T) {
...
+		{args: wrapArgs(ComplexType, "+17.59123j"), want: NewComplex(complex(0, 17.59123)).ToObject()},
+		{args: wrapArgs(ComplexType, "-3.0007e3j"), want: NewComplex(complex(0, -3000.7)).ToObject()},
+		{args: wrapArgs(ComplexType, "1+2j"), want: NewComplex(complex(1, 2)).ToObject()},
+		{args: wrapArgs(ComplexType, "3.1415-23j"), want: NewComplex(complex(3.1415, -23)).ToObject()},
+		{args: wrapArgs(ComplexType, "-23+3.1415j"), want: NewComplex(complex(-23, 3.1415)).ToObject()},
+		{args: wrapArgs(ComplexType, "+451.2192+384.27j"), want: NewComplex(complex(451.2192, 384.27)).ToObject()},
+		{args: wrapArgs(ComplexType, "-38.378-283.28j"), want: NewComplex(complex(-38.378, -283.28)).ToObject()},
+		{args: wrapArgs(ComplexType, "1.76123e2+0.000007e6j"), want: NewComplex(complex(176.123, 7)).ToObject()},
+		{args: wrapArgs(ComplexType, "-nan+nanj"), want: NewComplex(complex(math.NaN(), math.NaN())).ToObject()},
+		{args: wrapArgs(ComplexType, "inf-infj"), want: NewComplex(complex(math.Inf(1), math.Inf(-1))).ToObject()},
```
Comment (trotterdylan) [inline runtime/complex_test.go]: Looks like this is the sole failing test. According to my testing in CPython, your expectations are correct so you might have to handle this edge case specially.

Comment (m4ns0ur) [inline runtime/complex_test.go]: @trotterdylan not that one, but [this is the test case](https://github.com/google/grumpy/pull/296/files/0b650ff65a4b71dd0953e14a353aca01c3d070b2#diff-463afa6704d82d9d40baa2e71e0c1f63R206)
and here is the reason:
```
cr := complex(math.Inf(1), 0)
ci := complex(math.Inf(-1), 0)
fmt.Println(cr + ci*1i)
// getting (NaN-Infi), we expect (+Inf-Infi).
```
should I revert that correction part (just like [CPython](https://github.com/python/cpython/blob/master/Objects/complexobject.c#L1068))?

Comment (trotterdylan) [inline runtime/complex_test.go]: That seems like strange behavior in Go complex arithmetic. I think the correct outcome can be had by using the expression complex(real(cr)-imag(ci), imag(cr)+real(ci))

Comment (trotterdylan) [inline runtime/complex_test.go]: Also worth adding a comment about why we're using this expression instead of the simpler one we had originally.

**Comment by m4ns0ur** [comment]
really happy that we have these _python compatibility_ tests 👍
- just implemented `complex.__neg__` for test cases like `assert complex('-j') == -1j`
- have to tune `repr()` for test cases like `assert repr(complex('nan-nanj')) == "(nan+nanj)"`, we already get `"(NaNNaNj)"` for this

**Comment by trotterdylan** [comment]
@m4ns0ur: about the "strange assertion", https://github.com/google/grumpy/issues/262 may be relevant

**Comment by m4ns0ur** [comment]
@trotterdylan yes, same issue here.
would you mind if I remove this assertion test case for now, to be resolved under #262 ?

**Comment by trotterdylan** [comment]
Sounds good to me.

**Comment by m4ns0ur** [comment]
@trotterdylan ready to review

**Comment by trotterdylan** [comment]
Fantastic! Thanks for all your hard work on this. Merging.
